File: lib_local/gallery/src/main/java/com/yanzhenjie/mediascanner/MediaScanner.java
package com.yanzhenjie.mediascanner;

import android.content.Context;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.webkit.MimeTypeMap;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 媒体文件扫描器
 * 作用：通知Android系统刷新相册/文件管理器
 */
public class MediaScanner implements MediaScannerConnection.MediaScannerConnectionClient {
    // 扫描计数
    private int scanCount = 0;
    // 当前正在扫描的文件路径
    private String[] currentScanPaths;
    // 系统扫描连接
    private final MediaScannerConnection mediaScanConnection;
    // 扫描任务队列（线程安全）
    private final Queue<String[]> taskQueue = new LinkedList<>();

    /**
     * 构造方法
     */
    public MediaScanner(Context context) {
        mediaScanConnection = new MediaScannerConnection(context.getApplicationContext(), this);
    }

    /**
     * 系统扫描连接成功
     */
    @Override
    public void onMediaScannerConnected() {
        if (currentScanPaths == null || currentScanPaths.length == 0) return;
        scanCount = 0;
        for (String filePath : currentScanPaths) {
            String extension = MimeTypeMap.getFileExtensionFromUrl(filePath);
            String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
            mediaScanConnection.scanFile(filePath, mimeType);
        }
    }

    /**
     * 单个文件扫描完成
     */
    @Override
    public void onScanCompleted(String path, Uri uri) {
        scanCount++;
        // 当前批次全部扫描完毕
        if (scanCount >= currentScanPaths.length) {
            mediaScanConnection.disconnect();
            scanCount = 0;
            // 继续执行下一个任务
            executeNextTask();
        }
    }

    /**
     * 是否正在扫描
     */
    public boolean isRunning() {
        return mediaScanConnection.isConnected();
    }

    /**
     * 扫描单个文件
     */
    public void scan(String filePath) {
        if (filePath == null || filePath.isEmpty()) return;
        scan(new String[]{filePath});
    }

    /**
     * 扫描文件列表
     */
    public void scan(List<String> filePaths) {
        if (filePaths == null || filePaths.isEmpty()) return;
        scan(filePaths.toArray(new String[0]));
    }

    /**
     * 扫描文件数组
     */
    public void scan(String[] filePaths) {
        if (filePaths == null || filePaths.length == 0) return;
        taskQueue.offer(filePaths);
        if (!isRunning()) {
            executeNextTask();
        }
    }

    /**
     * 执行下一个扫描任务
     */
    private void executeNextTask() {
        // 队列为空，直接结束
        if (taskQueue.isEmpty()) {
            if (isRunning()) {
                mediaScanConnection.disconnect();
            }
            return;
        }
        // 取出并移除队首任务
        currentScanPaths = taskQueue.poll();
        if (currentScanPaths != null && currentScanPaths.length > 0) {
            mediaScanConnection.connect();
        }
    }

}

File: lib_local/gallery/src/main/java/com/yanzhenjie/mediascanner/MediaScanner.kt
package com.yanzhenjie.mediascanner

import android.content.Context
import android.media.MediaScannerConnection
import android.media.MediaScannerConnection.MediaScannerConnectionClient
import android.net.Uri
import android.webkit.MimeTypeMap
import java.util.LinkedList

/**
 * 将应用下载或生成的媒体文件（如图片、视频、音频）主动通知给系统的媒体库
 * 1) 对于 Android 13 (API 33) 及以上，需要 READ_MEDIA_IMAGES、READ_MEDIA_VIDEO、READ_MEDIA_AUDIO 等权限。
 * 2) 对于旧版本 Android，需要 READ_EXTERNAL_STORAGE 和 WRITE_EXTERNAL_STORAGE 权限
 * Created by yan
 */
class MediaScanner(context: Context) : MediaScannerConnectionClient {
    private var scanCount = 0
    private var mCurrentScanPaths = arrayOf<String>()
    private val mLinkedList = LinkedList<Array<String>>()
    private val mMediaScanConn = MediaScannerConnection(context.applicationContext, this)

    /**
     * 判断媒体扫描器当前是否正在运行（即是否处于连接状态）
     * true 表示扫描器正在运行，false 表示当前没有扫描任务。
     */
    fun isRunning(): Boolean {
        return mMediaScanConn.isConnected
    }

    /**
     * 提交文件路径进行扫描
     */
    fun scan(filePath: String?) {
        filePath ?: return
        scan(arrayOf(filePath))
    }

    fun scan(filePaths: MutableList<String>?) {
        filePaths ?: return
        scan(filePaths.toTypedArray<String>())
    }

    fun scan(filePaths: Array<String>?) {
        filePaths ?: return
        mLinkedList.add(filePaths)
        if (!isRunning()) {
            executeOnce()
        }
    }

    /**
     * 从任务队列中取出下一个任务，并开始执行（即开始扫描）
     */
    private fun executeOnce() {
        mCurrentScanPaths = mLinkedList[0]
        if (mCurrentScanPaths.isNotEmpty()) {
            mMediaScanConn.connect()
        }
    }

    /**
     * 成功与系统扫描服务建立连接后，该方法会被调用
     */
    override fun onMediaScannerConnected() {
        if (mCurrentScanPaths.isNotEmpty()) {
            for (filePath in mCurrentScanPaths) {
                val extension = MimeTypeMap.getFileExtensionFromUrl(filePath)
                val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
                mMediaScanConn.scanFile(filePath, mimeType)
            }
        }
    }

    /**
     * 当 scanFile() 请求的单个文件被成功扫描并添加到媒体库后，该方法会被调用
     */
    override fun onScanCompleted(path: String?, uri: Uri?) {
        scanCount++
        if (scanCount == mCurrentScanPaths.size) {
            mMediaScanConn.disconnect()
            scanCount = 0
            mLinkedList.poll()
            executeOnce()
        }
    }

}